solver/llbsolver/proc/provenance.go

```diff
@@ -0,0 +1,237 @@
+	"github.com/moby/buildkit/worker"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+var BuildKitBuildType = "https://mobyproject.org/buildkit@v1"
+
+func ProvenanceProcessor(attrs map[string]string) llbsolver.Processor {
+	return func(ctx context.Context, res *frontend.Result, s *llbsolver.Solver, j *solver.Job) (*frontend.Result, error) {
+		if len(res.Refs) == 0 {
+			return nil, errors.New("provided result has no refs")
+		}
+
+		platformsBytes, ok := res.Metadata[exptypes.ExporterPlatformsKey]
+		if !ok {
+			return nil, errors.Errorf("unable to collect multiple refs, missing platforms mapping")
+		}
+
+		var ps exptypes.Platforms
+		if len(platformsBytes) > 0 {
+			if err := json.Unmarshal(platformsBytes, &ps); err != nil {
+				return nil, errors.Wrapf(err, "failed to parse platforms passed to sbom processor")
+			}
+		}
+
+		buildID := identity.NewID()
+
+		var reproducible bool
+		if v, ok := attrs["reproducible"]; ok {
+			b, err := strconv.ParseBool(v)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to parse reproducible flag %q", v)
+			}
+			reproducible = b
+		}
+
+		var mode string
+		if v, ok := attrs["mode"]; ok {
+			switch v {
+			case "disabled", "none":
+				return res, nil
+			case "full":
+				mode = "max"
+			case "max", "min":
+				mode = v
+			default:
+				return nil, errors.Errorf("invalid mode %q", v)
+			}
+		}
+
+		for _, p := range ps.Platforms {
+			dt, ok := res.Metadata[exptypes.ExporterBuildInfo+"/"+p.ID]
+			if !ok {
+				return nil, errors.New("no build info found for provenance")
+			}
+
+			var bi binfotypes.BuildInfo
+			if err := json.Unmarshal(dt, &bi); err != nil {
+				return nil, errors.Wrap(err, "failed to parse build info")
+			}
+
+			pr, err := provenance.FromBuildInfo(bi)
+			if err != nil {
+				return nil, err
+			}
+
+			st := j.StartedTime()
+
+			pr.Metadata.BuildStartedOn = &st
+			pr.Metadata.Reproducible = reproducible
+			pr.Metadata.BuildInvocationID = buildID
+
+			var addLayers func() error
+
+			if mode != "max" {
+				param := make(map[string]*string)
+				for k, v := range pr.Invocation.Parameters.(map[string]*string) {
+					if strings.HasPrefix(k, "build-arg:") || strings.HasPrefix(k, "label:") {
+						pr.Metadata.Completeness.Parameters = false
+						continue
+					}
+					param[k] = v
+				}
+				pr.Invocation.Parameters = param
+			} else {
+				dgsts, err := provenance.AddBuildConfig(ctx, pr, res.Refs[p.ID])
+				if err != nil {
+					return nil, err
+				}
+
+				r, err := res.Refs[p.ID].Result(ctx)
+				if err != nil {
+					return nil, err
+				}
+
+				addLayers = func() error {
+					e := newCacheExporter()
+					if _, err := r.CacheKeys()[0].Exporter.ExportTo(ctx, e, solver.CacheExportOpt{
+						ResolveRemotes: resolveRemotes,
+						Mode:           solver.CacheExportModeRemoteOnly,
+						ExportRoots:    true,
+					}); err != nil {
+						return err
+					}
```

Comment:
So I'm a little hazy on the details of how this works sorry! This seems like we are able to calculate the digests of layers before we export? How is this handled if we are exporting `local` or `tar` where we don't have layer identifiers? Why doesn't this have similar problems for calculating layer ids for passing to an SBOM scanner (at least for the post-build proc scanner)?

Reply:
> This seems like we are able to calculate the digests of layers before we export?

No, this is after image has created layers. That's why it is called inside the callback. If there are no layers then the info for that part will not be available.

In SBOM scanner you already have the refs that the scanner ran on. You just need to find files from these layers.

In here we don't care about the files, but we want to find all the layers that participated in this build, not only the layer in the final image. Cache exporting walks back all the build steps and exports all the layers it knows about in min/remote-only mode (even if layer isn't available locally for example but only in cache-from source or somewhere in a registry). We can reuse the same code in here. The only part that was missing is how to associate the export location with the LLB digest of the specific build step.